intermediate_source/forward_ad_tutorial.py

```diff
@@ -0,0 +1,167 @@
+# Usage with Modules
+# --------------------------------------------------------------------
+# To use ``nn.Module``s with forward AD, replace the parameters of your
+# model with dual tensors before performing the forward pass. At the
+# time of writing, it is not possible to create dual tensor
+# `nn.Parameter`s. As a workaround, one must register the dual tensor
+# as a non-parameter attribute of the module.
+
+import torch.nn as nn
+
+model = nn.Linear(5, 5)
+input = torch.randn(16, 5)
+
+params = {name: p for name, p in model.named_parameters()}
+tangents = {name: torch.rand_like(p) for name, p in params.items()}
+
+with fwAD.dual_level():
+    for name, p in params.items():
+        delattr(model, name)
+        setattr(model, name, fwAD.make_dual(p, tangents[name]))
+
+    out = model(input)
+    jvp = fwAD.unpack_dual(out).tangent
+
+######################################################################
+# Using Modules stateless API (experimental)
+# --------------------------------------------------------------------
+# Another way to use ``nn.Module``s with forward AD is to utilize
+# the stateless API. NB: At the time of writing the stateless API is still
+# experimental and may be subject to change.
+
+from torch.nn.utils._stateless import functional_call
+
+# We need a fresh module because the functional call requires the
+# the model to have parameters registered.
+model = nn.Linear(5, 5)
+
+dual_params = {}
+with fwAD.dual_level():
+    for name, p in params.items():
+        # Using the same ``tangents`` from the above section
+        dual_params[name] = fwAD.make_dual(p, tangents[name])
+    out = functional_call(model, dual_params, input)
+    jvp2 = fwAD.unpack_dual(out).tangent
+
+# Check our results
+assert torch.allclose(jvp, jvp2)
+
+######################################################################
+# Custom autograd Function
+# --------------------------------------------------------------------
+# Custom Functions also support forward-mode AD. To create custom Function
+# supporting forward-mode AD, register the ``jvp()`` static method. It is
+# possible, but not mandatory for custom Functions to support both forward
+# and backward AD. See the
+# `documentation <https://pytorch.org/docs/master/notes/extending.html#forward-mode-ad>`_
+# for more information.
+
+class Fn(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        result = torch.exp(foo)
+        # Tensors stored in ctx can be used in the subsequent forward grad
+        # computation.
+        ctx.result = result
+        return result
+
+    @staticmethod
+    def jvp(ctx, gI):
+        gO = gI * ctx.result
+        # If the tensor stored in ctx will not also be used in the backward pass,
+        # one can manually free it using ``del``
+        del ctx.result
+        return gO
+
+fn = Fn.apply
+
+primal = torch.randn(10, 10, dtype=torch.double, requires_grad=True)  # Fix this?
```

Comment:
Gradcheck should not require inputs to require grad if we are not checking backward AD. See:  https://github.com/pytorch/pytorch/pull/69004

Reply:
Not sure this is true.
We do use requires_grad to know which input/outputs are differentiable.

Maybe we should not, but today we do require this, and just removing the check doesn't sound like the right solution.

Reply:
You're correct about the inputs, but looks like gradcheck currently doesn't actually need the outputs to have requires grad:
```python
import torch
import torch.autograd.forward_ad as fwAD

def detach_backward(x):
  _, tangent = fwAD.unpack_dual(x)
  if not tangent:   # non-dual-level friendly path
    return x.detach()
  return fwAD.make_dual(x.detach(), tangent)

class Func(torch.autograd.Function):
  @staticmethod
  def forward(ctx, x):
    return x * 2

  @staticmethod
  def jvp(ctx, gI):
    return gI * 3  # Wrong, lets see if gradcheck catches this

t = torch.tensor(1., requires_grad=True, dtype=torch.double)

def fn(x):
  out1 = detach_backward(Func.apply(x))
  out2 = x.clone()  # only necessary to bypass the no differnetiable outputs check
  return out1, out2

with fwAD.dual_level():
  dual = fwAD.make_dual(t, torch.rand_like(t))
  out = fn(dual)
  assert not out[0].requires_grad

torch.autograd.gradcheck(fn, (t,), check_forward_ad=True,
  check_backward_ad=False, check_batched_grad=False, check_undefined_grad=False)
```

I don't know whether the fact non-require-grad outputs are checked is useful though. At least for custom functions, it looks like the output still requires_grad even if only `jvp` is implemented, i..e, does a function that behaves like `detach_backward` actually exist in nature?

Reply:
> does a function that behaves like detach_backward actually exist in nature?

Ignoring things like `detach`, I think the answer is no. If a function is differentiable, it will be for both forward and backward.